File: services/api-gateway/app/schemas/pharmacy.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


# ── Pharmacy Item / Inventory schemas ──────────────────────────────────────


class PharmacyItemCreate(BaseModel):
    """Schema for adding a drug to pharmacy inventory."""

    drug_code: str = Field(..., min_length=1, max_length=50)
    drug_name: str = Field(..., min_length=1, max_length=300)
    generic_name: str | None = Field(None, max_length=300)
    atc_code: str | None = Field(None, max_length=20)
    is_keml: bool = False
    dosage_form: str | None = Field(
        None, pattern=r"^(tablet|capsule|syrup|suspension|injection|cream|ointment|drops|inhaler|suppository|sachet|solution|powder)$"
    )
    strength: str | None = Field(None, max_length=50)
    batch_number: str | None = Field(None, max_length=50)
    current_quantity: int = Field(0, ge=0)
    unit_of_measure: str = Field(
        "tablet",
        pattern=r"^(tablet|capsule|ml|vial|tube|sachet|bottle|ampoule|dose)$",
    )
    reorder_level: int = Field(10, ge=0)
    max_stock_level: int | None = Field(None, ge=0)
    buying_price_cents: int | None = Field(None, ge=0)
    selling_price_cents: int | None = Field(None, ge=0)
    expiry_date: date | None = None
    manufacturer: str | None = Field(None, max_length=200)
    supplier: str | None = Field(None, max_length=200)
    storage_conditions: str | None = Field(None, max_length=100)
    shelf_location: str | None = Field(None, max_length=50)


class PharmacyItemUpdate(BaseModel):
    """Schema for updating a pharmacy inventory item."""

    drug_name: str | None = Field(None, min_length=1, max_length=300)
    generic_name: str | None = Field(None, max_length=300)
    is_keml: bool | None = None
    current_quantity: int | None = Field(None, ge=0)
    reorder_level: int | None = Field(None, ge=0)
    selling_price_cents: int | None = Field(None, ge=0)
    expiry_date: date | None = None
    shelf_location: str | None = Field(None, max_length=50)
    is_active: bool | None = None


class PharmacyItemResponse(BaseModel):
    """Schema for pharmacy item API responses."""

    id: uuid.UUID
    facility_id: uuid.UUID
    drug_code: str
    drug_name: str
    generic_name: str | None
    atc_code: str | None
    is_keml: bool
    dosage_form: str | None
    strength: str | None
    batch_number: str | None
    current_quantity: int
    unit_of_measure: str
    reorder_level: int
    max_stock_level: int | None
    buying_price_cents: int | None
    selling_price_cents: int | None
    expiry_date: date | None
    manufacturer: str | None
    supplier: str | None
    storage_conditions: str | None
    shelf_location: str | None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class InventoryListResponse(BaseModel):
    """Paginated pharmacy inventory list."""

    items: list[PharmacyItemResponse]
    total: int
    page: int
    page_size: int


class StockAlert(BaseModel):
    """Low stock or expiry alert."""

    item_id: uuid.UUID
    drug_name: str
    drug_code: str
    current_quantity: int
    reorder_level: int
    expiry_date: date | None
    alert_type: str  # low_stock, expiring_soon, expired, out_of_stock


# ── Dispensing schemas ─────────────────────────────────────────────────────


class DispenseRequest(BaseModel):
    """Schema for dispensing a prescription."""

    prescription_id: uuid.UUID
    patient_id: uuid.UUID
    pharmacy_item_id: uuid.UUID | None = None
    quantity_dispensed: int = Field(..., ge=1)
    substitution_drug: str | None = Field(None, max_length=300)
    substitution_reason: str | None = Field(None, max_length=500)
    dosage_instructions: str | None = Field(None, max_length=1000)
    counseling_done: bool = False
    counseling_notes: str | None = Field(None, max_length=2000)
    payment_method: str | None = Field(
        None, pattern=r"^(cash|mpesa|insurance|exemption)$"
    )


class DispensingResponse(BaseModel):
    """Schema for dispensing API responses."""

    id: uuid.UUID
    prescription_id: uuid.UUID
    patient_id: uuid.UUID
    pharmacy_item_id: uuid.UUID | None
    dispensed_by: uuid.UUID
    drug_name: str
    quantity_requested: int
    quantity_dispensed: int
    substitution_drug: str | None
    substitution_reason: str | None
    dosage_instructions: str | None
    counseling_done: bool
    counseling_notes: str | None
    status: str
    unit_price_cents: int | None
    total_price_cents: int | None
    payment_status: str
    payment_method: str | None
    dispensed_at: datetime
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Pharmacy Queue schemas ────────────────────────────────────────────────


class PharmacyQueueItem(BaseModel):
    """A pending prescription in the pharmacy queue."""

    prescription_id: uuid.UUID
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    patient_name: str | None = None
    patient_mrn: str | None = None
    drug_name: str
    generic_name: str | None
    is_keml: bool
    dosage: str
    route: str
    frequency: str
    duration_days: int | None
    quantity: int | None
    instructions: str | None
    prescriber_id: uuid.UUID
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class PharmacyQueueResponse(BaseModel):
    """Pharmacy queue list response."""

    items: list[PharmacyQueueItem]
    total: int


# ── Stock Transaction schemas ─────────────────────────────────────────────


class StockReceiptRequest(BaseModel):
    """Schema for receiving new stock."""

    pharmacy_item_id: uuid.UUID
    quantity: int = Field(..., ge=1)
    batch_number: str | None = Field(None, max_length=50)
    expiry_date: date | None = None
    unit_cost_cents: int | None = Field(None, ge=0)
    reference_number: str | None = Field(None, max_length=100)
    reason: str | None = Field(None, max_length=500)


class StockAdjustmentRequest(BaseModel):
    """Schema for stock adjustments (loss, damage, transfer, correction)."""

    pharmacy_item_id: uuid.UUID
    quantity_change: int  # positive for add, negative for deduct
    reason: str = Field(..., min_length=1, max_length=500)
    reference_number: str | None = Field(None, max_length=100)


class StockTransactionResponse(BaseModel):
    """Schema for stock transaction API responses."""

    id: uuid.UUID
    pharmacy_item_id: uuid.UUID
    dispensing_id: uuid.UUID | None
    transaction_type: str
    quantity_change: int
    quantity_before: int
    quantity_after: int
    reference_number: str | None
    reason: str | None
    unit_cost_cents: int | None
    batch_number: str | None
    expiry_date: date | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: packages/shared/src/types/pharmacy.ts
/** Pharmacy inventory item. */
export interface PharmacyItem {
  id: string;
  facility_id: string;
  drug_code: string;
  drug_name: string;
  generic_name: string | null;
  atc_code: string | null;
  is_keml: boolean;
  dosage_form: string | null;
  strength: string | null;
  batch_number: string | null;
  current_quantity: number;
  unit_of_measure: string;
  reorder_level: number;
  max_stock_level: number | null;
  buying_price_cents: number | null;
  selling_price_cents: number | null;
  expiry_date: string | null;
  manufacturer: string | null;
  supplier: string | null;
  storage_conditions: string | null;
  shelf_location: string | null;
  is_active: boolean;
  created_at: string;
}

/** Payload for creating a pharmacy item. */
export interface PharmacyItemCreate {
  drug_code: string;
  drug_name: string;
  generic_name?: string | null;
  atc_code?: string | null;
  is_keml?: boolean;
  dosage_form?: string | null;
  strength?: string | null;
  batch_number?: string | null;
  current_quantity?: number;
  unit_of_measure?: string;
  reorder_level?: number;
  max_stock_level?: number | null;
  buying_price_cents?: number | null;
  selling_price_cents?: number | null;
  expiry_date?: string | null;
  manufacturer?: string | null;
  supplier?: string | null;
  storage_conditions?: string | null;
  shelf_location?: string | null;
}

/** Payload for updating a pharmacy item. */
export interface PharmacyItemUpdate {
  drug_name?: string;
  generic_name?: string | null;
  is_keml?: boolean;
  current_quantity?: number;
  reorder_level?: number;
  selling_price_cents?: number | null;
  expiry_date?: string | null;
  shelf_location?: string | null;
  is_active?: boolean;
}

/** Paginated inventory list response. */
export interface InventoryListResponse {
  items: PharmacyItem[];
  total: number;
  page: number;
  page_size: number;
}

/** A pending prescription in the pharmacy queue. */
export interface PharmacyQueueItem {
  prescription_id: string;
  encounter_id: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  drug_name: string;
  generic_name: string | null;
  is_keml: boolean;
  dosage: string;
  route: string;
  frequency: string;
  duration_days: number | null;
  quantity: number | null;
  instructions: string | null;
  prescriber_id: string;
  status: string;
  created_at: string;
}

/** Pharmacy queue response. */
export interface PharmacyQueueResponse {
  items: PharmacyQueueItem[];
  total: number;
}

/** Dispensing record. */
export interface Dispensing {
  id: string;
  prescription_id: string;
  patient_id: string;
  pharmacy_item_id: string | null;
  dispensed_by: string;
  drug_name: string;
  quantity_requested: number;
  quantity_dispensed: number;
  substitution_drug: string | null;
  substitution_reason: string | null;
  dosage_instructions: string | null;
  counseling_done: boolean;
  counseling_notes: string | null;
  status: string;
  unit_price_cents: number | null;
  total_price_cents: number | null;
  payment_status: string;
  payment_method: string | null;
  dispensed_at: string;
  created_at: string;
}

/** Payload for dispensing a prescription. */
export interface DispenseRequest {
  prescription_id: string;
  patient_id: string;
  pharmacy_item_id?: string | null;
  quantity_dispensed: number;
  substitution_drug?: string | null;
  substitution_reason?: string | null;
  dosage_instructions?: string | null;
  counseling_done?: boolean;
  counseling_notes?: string | null;
  payment_method?: string | null;
}

/** Stock alert (low stock, expiry, etc.). */
export interface StockAlert {
  item_id: string;
  drug_name: string;
  drug_code: string;
  current_quantity: number;
  reorder_level: number;
  expiry_date: string | null;
  alert_type: "low_stock" | "expiring_soon" | "expired" | "out_of_stock";
}

/** Payload for receiving stock. */
export interface StockReceiptRequest {
  pharmacy_item_id: string;
  quantity: number;
  batch_number?: string | null;
  expiry_date?: string | null;
  unit_cost_cents?: number | null;
  reference_number?: string | null;
  reason?: string | null;
}

/** Payload for stock adjustment. */
export interface StockAdjustmentRequest {
  pharmacy_item_id: string;
  quantity_change: number;
  reason: string;
  reference_number?: string | null;
}

/** Stock transaction record. */
export interface StockTransaction {
  id: string;
  pharmacy_item_id: string;
  dispensing_id: string | null;
  transaction_type: string;
  quantity_change: number;
  quantity_before: number;
  quantity_after: number;
  reference_number: string | null;
  reason: string | null;
  unit_cost_cents: number | null;
  batch_number: string | null;
  expiry_date: string | null;
  created_at: string;
}

/**
 * Payload for topping up an existing pharmacy item (URL-path variant).
 *
 * Used with `POST /pharmacy/inventory/{item_id}/receive` so callers cannot
 * create a duplicate catalogue row when restocking.
 */
export interface InventoryReceiveRequest {
  quantity: number;
  batch_no?: string | null;
  expiry_date?: string | null;
  supplier?: string | null;
  /** Per-unit cost in KES (decimal string or number). */
  unit_cost?: string | number | null;
  notes?: string | null;
  reference_number?: string | null;
}

/** Payload for stock-take corrections — sets stock to `new_qty` exactly. */
export interface InventoryAdjustRequest {
  new_qty: number;
  reason: string;
}

/** Single row of the Opening / Received / Sales / Closing stock summary. */
export interface StockSummaryRow {
  item_id: string;
  item_name: string;
  drug_code: string;
  unit: string;
  opening_stock: number;
  received_stock: number;
  sales_stock: number;
  closing_stock: number;
  current_stock: number;
}

/** Stock summary response for a date range. */
export interface StockSummaryResponse {
  start_date: string;
  end_date: string;
  rows: StockSummaryRow[];
  total_items: number;
}
